### ISSUE TITLE
Analytics: Declarative event accumulation engine with PreboundApi-style registration

### ISSUE THREAD

**Issue by DonLoquacious** [beyond-immersion/bannou-service#703]
## Summary

Design and implement a declarative event accumulation engine in Analytics (L4) where services register "accumulation templates" ahead of time, Analytics processes matching events through those templates and maintains accumulators, and any service can bulk-query accumulated data by category. The registration reuses the existing PreboundApi / ResponseTransformation infrastructure (JsonPath extraction, conditional filtering) from the Core SDK.

**Motivation**: Multiple systems need "count of things that happened" — Currency needs TotalMinted/TotalBurned, Quest needs kill counts, Divine needs domain activity metrics, leveling needs XP tracking. Analytics is the correct home for this (L4 optional observer), but it currently has no generic accumulation infrastructure. The [DIVINITY-GENERATION-ARCHITECTURE.md](https://github.com/beyond-immersion/bannou-service/blob/master/docs/planning/DIVINITY-GENERATION-ARCHITECTURE.md) established the principle: "Analytics is the most optional plugin — it must NEVER be in the critical path for game mechanics." This engine makes Analytics useful for enrichment data (TotalMinted/TotalBurned, economic dashboards) while keeping it fully optional — consumers get `null` when Analytics is absent.

**First consumer**: Currency's `GetGlobalSupply` endpoint (#211) needs `TotalMinted` and `TotalBurned` values. These are game-design-dependent classifications of transaction types — exactly what this engine is for.

---

### DI Provider Inversion (Two Interfaces in `bannou-service/Providers/`)

Services can't call Analytics (L4) directly (hierarchy violation for L1/L2). Two DI interfaces handle both directions:

| Interface | Direction | Purpose |
|---|---|---|
| `IAnalyticsAccumulationRegistrant` | L1/L2/L4 → L4 (discovery) | Services declare accumulation templates; Analytics discovers at startup via `IEnumerable<T>` |
| `IAccumulatedDataProvider` | L4 → L1/L2 (pull) | Services query accumulated data; Analytics implements; consumers discover via `IEnumerable<T>` |

Follows established patterns: `IPermissionRegistrant` (discovery), `IVariableProviderFactory` (pull).

### Accumulation Template Model

Each registration returns one or more templates:

```
AccumulationTemplate
├── category          string ("currency", "quest", "divine", "combat")
├── fieldName         string ("totalMinted", "totalBurned", "killCount")
├── eventTopic        string ("currency.credited", "currency.debited")
├── groupByPath       string? (JsonPath for per-entity partitioning, e.g., "$.currencyDefinitionId")
├── accumulationMode  enum: Sum | Count | Min | Max | Latest
├── valuePath         string (JsonPath to extract the numeric value, e.g., "$.amount")
├── conditions        List<TransformationCondition>? (reuse from Core SDK — filter events)
└── description       string? (human-readable for diagnostics)
```

**Example — Currency TotalMinted registration:**

```
category: "currency"
fieldName: "totalMinted"
eventTopic: "currency.credited"
groupByPath: "$.currencyDefinitionId"
accumulationMode: Sum
valuePath: "$.amount"
conditions:
  - type: JsonPathEquals, jsonPath: "$.transactionType", expectedValue: "Mint"
  OR
  - type: JsonPathEquals, jsonPath: "$.transactionType", expectedValue: "QuestReward"
  OR
  - type: JsonPathEquals, jsonPath: "$.transactionType", expectedValue: "LootDrop"
  ... etc.
description: "Sum of all faucet credits per currency definition"
```

**Note on conditions**: The current `TransformationCondition` model uses AND logic within a rule. For OR-style filtering (multiple acceptable TransactionType values), either: (a) add a `JsonPathIn` condition type that matches against a list of values, or (b) use the existing `JsonPathContains` with a comma-delimited value list convention. Option (a) is cleaner and a natural extension of the condition type enum.

### Bulk-Get Query Interface

```csharp
// IAccumulatedDataProvider (implemented by Analytics, discovered by consumers)
Task<AccumulatedDataResult?> GetAccumulatedAsync(
    string category,
    string? groupKey,           // e.g., currencyDefinitionId.ToString()
    string[] fields,            // e.g., ["totalMinted", "totalBurned"]
    CancellationToken ct);

// Also: bulk variant for multiple group keys in one call
Task<Dictionary<string, AccumulatedDataResult>?> GetAccumulatedBatchAsync(
    string category,
    string[] groupKeys,
    string[] fields,
    CancellationToken ct);
```

```
AccumulatedDataResult
├── values          Dictionary<string, double> (fieldName → accumulated value)
├── lastUpdatedAt   DateTimeOffset (when the accumulator was last flushed to persistent storage)
```

Currency's `GetGlobalSupply` then does:
```csharp
var providers = _accumulatedDataProviders; // IEnumerable<IAccumulatedDataProvider>
double? totalMinted = null, totalBurned = null;
foreach (var provider in providers)
{
    var result = await provider.GetAccumulatedAsync(
        "currency", body.CurrencyDefinitionId.ToString(),
        new[] { "totalMinted", "totalBurned" }, ct);
    if (result != null)
    {
        result.Values.TryGetValue("totalMinted", out var m); totalMinted = m;
        result.Values.TryGetValue("totalBurned", out var b); totalBurned = b;
    }
}
```

No Analytics? No provider registered. `totalMinted = null`, `totalBurned = null`. Fully optional.

---

## Multi-Node Analytics: Topic Distribution

Analytics subscribes to many (potentially all) service event topics. In a multi-node Analytics deployment, naive fanout means every node processes every event — double/triple-counting accumulators.

### Design: Partitioned Consumption

Analytics should use **partitioned topic consumption** rather than fanout for accumulation work:

- **Single-consumer queue per topic** for accumulation (not the shared `IEventConsumer` fanout path). Each accumulation topic gets its own dedicated RabbitMQ queue with a single active consumer.
- **Distribution across nodes**: When multiple Analytics nodes are present, topics are distributed across them — each topic's accumulation queue is consumed by exactly one node. If a node dies, RabbitMQ redelivers to a surviving node (standard competing-consumer semantics on a named queue).
- **Topic assignment**: At startup, Analytics nodes coordinate topic distribution. Options:
  - **Static partitioning**: Hash topic name → node assignment (simple, no coordination)
  - **Dynamic rebalancing**: Leader election or consensus to distribute topics evenly (more complex but handles node join/leave gracefully)
  - **RabbitMQ native**: Use `x-single-active-consumer` queue flag (RabbitMQ 3.8+) — declares that only one consumer should be active at a time. Combined with per-topic named queues, this gives exactly-once-per-topic processing with automatic failover.

**Recommendation**: `x-single-active-consumer` on per-topic named queues is the simplest correct solution. Each Analytics node declares consumers for all accumulation topics. RabbitMQ ensures only one node's consumer is active per queue. Failover is automatic. No custom coordination needed.

### Separation of Concerns

Accumulation (writes to counters) uses partitioned single-consumer queues. Query serving (reads from counters) runs on all nodes — read path is eventually consistent, no coordination needed.

---

### Design: EventBatcher Pattern for Accumulation

Each Analytics node maintains in-memory accumulators using the established `EventBatcher<T>` / `ConcurrentDictionary` pattern. Events are processed into in-memory counters immediately (microsecond-fast). Counters are flushed periodically to MySQL (as JSON blobs per category+groupKey) for persistence across restarts.

```
Event arrives (single-consumer queue)
    ↓
In-memory ConcurrentDictionary<(category, groupKey, field), double> accumulator
    ↓ (immediate, lock-free atomic update)
    
AccumulationFlushWorker (periodic, every AccumulationFlushIntervalSeconds)
    ↓
Drain accumulator atomically (Interlocked.Exchange pattern from EventBatcher)
    ↓
Group by (category, groupKey) → JSON blob { field: value, field: value, ... }
    ↓
WRITE to MySQL: analytics-accumulations store, key = "accum:{category}:{groupKey}"
    ↓
Values are merged (Sum: +=, Count: +=, Min: min, Max: max, Latest: replace)
```

### Multi-Node Consistency

Each node accumulates events from its assigned topics (partitioned, no overlap). When flushing to MySQL:
- Each node writes its OWN accumulated deltas (not absolute values)
- MySQL merge uses atomic read-modify-write (ETag concurrency) to add deltas to stored totals
- Result: all nodes contribute their portion, MySQL holds the global total
- Query from any node reads from MySQL → eventually consistent (staleness = max flush interval)

### Restart Recovery

On startup, Analytics reads existing accumulation values from MySQL. In-memory accumulators initialize from the persisted values. No event replay needed. Maximum data loss = one flush interval worth of events (configurable, e.g., 10-30 seconds).

### Configuration

| Property | Default | Purpose |
|---|---|---|
| `AccumulationFlushIntervalSeconds` | `10` | How often in-memory accumulators flush to MySQL |
| `AccumulationFlushStartupDelaySeconds` | `15` | Delay before first flush after startup |

---

## Accumulation Modes

| Mode | Operation | Merge (multi-node) | Use Case |
|---|---|---|---|
| `Sum` | `+= value` | `+= delta` | TotalMinted, TotalBurned, XP earned |
| `Count` | `+= 1` | `+= delta` | Number of transactions, kills |
| `Min` | `min(current, value)` | `min(stored, delta)` | Lowest single credit |
| `Max` | `max(current, value)` | `max(stored, delta)` | Largest single credit |
| `Latest` | `= value` | `= value` (last-write-wins) | Most recent exchange rate |

`Sum` and `Count` are additively mergeable across nodes. `Min`/`Max` are mergeable via their own operators. `Latest` is last-write-wins.

---

## Schema Drift Safety

Accumulation templates reference event fields by JsonPath string. If an event schema renames a field, the template silently stops matching.

**Mitigation**: Structural test in `structural-tests/` that validates all registered accumulation templates against event schemas at build time:
- For each `IAnalyticsAccumulationRegistrant` implementation found via reflection
- Extract all templates
- Resolve the `eventTopic` to its event schema type
- Validate that `valuePath`, `groupByPath`, and all condition `JsonPath` values resolve to existing fields in the event type

**Transition handling for field renames**: The registrant's responsibility. During a transition period, register templates for BOTH the old and new field names. The registrant does the math (sum old + new values) when queried. The structural test allows both to coexist temporarily.

---

## Condition Extension: JsonPathIn

The current `TransformationCondition` model uses AND logic within a rule. For accumulation filtering (e.g., "TransactionType is one of [Mint, QuestReward, LootDrop, ...]"), a `JsonPathIn` condition type is needed:

```csharp
/// <summary>JSON path value matches one of the expected values (case-insensitive, OR logic).</summary>
JsonPathIn
```

With a corresponding `ExpectedValues` (list of strings) property on `TransformationCondition`. This is a natural extension of the condition type enum in Core SDK.

---

## Related Issues

- **#211** (Currency: Global Supply Analytics) — first consumer; TotalMinted/TotalBurned fields
- **#429** (Analytics: Economic Velocity & Distribution Extensions) — broader Analytics extensions
- **#636** (Divine: domain-to-analytics mapping) — dissolved by DIVINITY-GENERATION-ARCHITECTURE; the accumulation engine would replace any residual need

## Related Documents

- [DIVINITY-GENERATION-ARCHITECTURE.md](https://github.com/beyond-immersion/bannou-service/blob/master/docs/planning/DIVINITY-GENERATION-ARCHITECTURE.md) — "Analytics is wrong for game mechanics" principle, wallet-as-accumulator pattern
- `sdks/core/PreboundApi.cs` — PreboundApi model with ResponseTransformation
- `sdks/core/ResponseTransformation.cs` — TransformationCondition, JsonPath conditions
- `sdks/core/ResponseTransformer.cs` — Pure-computation transformation engine
